File: static/kinematic.js
/**
 * Kinematic Scene Renderer for Three.js
 * Only renders - all calculations done on server
 */

class SceneNode {
    constructor(name, modelData) {
        this.name = name;
        this.group = new THREE.Group();
        this.group.name = name;
        this.mesh = null;

        if (modelData) {
            this.createMesh(modelData);
        }
    }

    createMesh(modelData) {
        const type = modelData.type || 'cube';
        let geometry;
        const material = new THREE.MeshPhongMaterial({
            color: modelData.color || 0x9e9e9e,
            flatShading: false
        });

        switch (type) {
            case 'cube':
                const size = modelData.size || [10, 10, 10];
                geometry = new THREE.BoxGeometry(size[0], size[1], size[2]);
                break;

            case 'sphere':
                const radius = modelData.radius || 10;
                geometry = new THREE.SphereGeometry(radius, 32, 32);
                break;

            case 'cylinder':
                const cylRadius = modelData.radius || 10;
                const cylHeight = modelData.height || 20;
                geometry = new THREE.CylinderGeometry(cylRadius, cylRadius, cylHeight, 32);
                break;

            case 'cone':
                const r1 = modelData.radius1 || 10;
                const r2 = modelData.radius2 || 5;
                const coneHeight = modelData.height || 20;
                geometry = new THREE.CylinderGeometry(r2, r1, coneHeight, 32);
                break;

            case 'file':
            case 'stl':
                this.loadExternalModel(modelData);
                // Placeholder while loading
                geometry = new THREE.BoxGeometry(1, 1, 1);
                material.visible = false;
                break;

            case 'none':
                // No visual representation for this node
                return;

            default:
                console.warn(`Unknown model type: ${type}, using cube fallback`);
                geometry = new THREE.BoxGeometry(10, 10, 10);
        }

        this.mesh = new THREE.Mesh(geometry, material);
        this.group.add(this.mesh);
    }

    loadExternalModel(modelData) {
        const path = modelData.path;
        if (!path) return;

        const url = path.startsWith('/') ? path : `/static/models/${path}`;
        const extension = path.split('.').pop().toLowerCase();

        if (extension === 'stl') {
            const loader = new THREE.STLLoader();
            loader.load(
                url,
                (geometry) => {
                    const material = new THREE.MeshPhongMaterial({
                        color: modelData.color || 0x9e9e9e,
                        flatShading: false
                    });
                    const mesh = new THREE.Mesh(geometry, material);

                    if (modelData.scale) {
                        if (Array.isArray(modelData.scale)) {
                            mesh.scale.set(modelData.scale[0], modelData.scale[1], modelData.scale[2]);
                        } else {
                            mesh.scale.setScalar(modelData.scale);
                        }
                    }

                    if (modelData.rotation) {
                        const r = modelData.rotation;
                        mesh.rotation.set(
                            r[0] * Math.PI / 180,
                            r[1] * Math.PI / 180,
                            r[2] * Math.PI / 180
                        );
                    }

                    if (modelData.offset) {
                        mesh.position.set(modelData.offset[0], modelData.offset[1], modelData.offset[2]);
                    }

                    if (this.mesh) {
                        this.group.remove(this.mesh);
                    }
                    this.mesh = mesh;
                    this.group.add(mesh);

                    console.log(`Loaded STL: ${path}`);
                },
                undefined,
                (error) => console.error(`Failed to load STL ${path}:`, error)
            );
        }
    }

    setPose(pose) {
        if (pose.position) {
            this.group.position.set(pose.position[0], pose.position[1], pose.position[2]);
        }
        if (pose.orientation) {
            this.group.quaternion.set(
                pose.orientation[0],
                pose.orientation[1],
                pose.orientation[2],
                pose.orientation[3]
            );
        }
    }
}


class KinematicScene {
    constructor(scene) {
        this.scene = scene;
        this.nodes = {};
        this.treeData = null;  // Original tree structure for local calculations
        this.zUp = false;

        // Root group for Z-up to Y-up conversion
        this.rootGroup = new THREE.Group();
        this.rootGroup.name = '__kinematic_root__';
        this.scene.add(this.rootGroup);
    }

    setZUp(enabled) {
        this.zUp = enabled;
        if (enabled) {
            // Rotate -90 degrees around X axis to convert Z-up to Y-up
            this.rootGroup.rotation.x = -Math.PI / 2;
        } else {
            this.rootGroup.rotation.x = 0;
        }
        console.log(`Z-up mode: ${enabled}`);
    }

    initFromSceneData(sceneData) {
        const nodeNames = Object.keys(sceneData);
        console.log(`initFromSceneData: creating ${nodeNames.length} nodes:`, nodeNames);

        // Create nodes from scene data
        for (const [name, data] of Object.entries(sceneData)) {
            if (!this.nodes[name]) {
                const node = new SceneNode(name, data.model);
                this.nodes[name] = node;
                this.rootGroup.add(node.group);  // Add to rootGroup for Z-up support
                console.log(`  Created: ${name}`);
            } else {
                console.log(`  Already exists: ${name}`);
            }
            // Apply initial pose
            if (data.pose) {
                this.nodes[name].setPose(data.pose);
            }
        }

        console.log(`Scene now has ${Object.keys(this.nodes).length} nodes`);
    }

    setTreeData(data) {
        this.treeData = data;
        this.jointCoords = {};
        // Initialize joint coords to 0
        this._collectJoints(data);
    }

    _collectJoints(node) {
        if (node.type === 'rotator' || node.type === 'actuator') {
            this.jointCoords[node.name] = 0;
        }
        for (const child of (node.children || [])) {
            this._collectJoints(child);
        }
    }

    updateFromSceneData(sceneData) {
        for (const [name, data] of Object.entries(sceneData)) {
            if (this.nodes[name] && data.pose) {
                this.nodes[name].setPose(data.pose);
            }
        }
    }

    // Local calculation methods
    setLocalJointCoord(name, value) {
        if (this.jointCoords.hasOwnProperty(name)) {
            this.jointCoords[name] = value;
        }
    }

    updateLocal() {
        if (!this.treeData) return;
        this._computePoses(this.treeData, { position: [0,0,0], orientation: [0,0,0,1] });
    }

    _computePoses(node, parentPose) {
        // Compose parent * local * joint
        // Joint transform is included so the node's visual moves/rotates with the joint
        const localPose = node.pose || { position: [0,0,0], orientation: [0,0,0,1] };
        let globalPose = this._composePoses(parentPose, localPose);

        // Apply joint transform to this node's pose (not just children)
        if (node.type === 'rotator' || node.type === 'actuator') {
            const jointTransform = this._getJointTransform(node);
            globalPose = this._composePoses(globalPose, jointTransform);
        }

        // Apply to scene node
        if (this.nodes[node.name]) {
            this.nodes[node.name].setPose(globalPose);
        }

        // Recurse children (they inherit parent's full transform including joint)
        for (const child of (node.children || [])) {
            this._computePoses(child, globalPose);
        }
    }

    _getJointTransform(node) {
        const coord = this.jointCoords[node.name] || 0;
        const axis = node.axis || [0, 0, 1];

        if (node.type === 'rotator') {
            const q = this._quatFromAxisAngle(axis, coord);
            return { position: [0, 0, 0], orientation: q };
        } else if (node.type === 'actuator') {
            return {
                position: [axis[0] * coord, axis[1] * coord, axis[2] * coord],
                orientation: [0, 0, 0, 1]
            };
        }
        return { position: [0, 0, 0], orientation: [0, 0, 0, 1] };
    }

    _quatFromAxisAngle(axis, angle) {
        const half = angle / 2;
        const s = Math.sin(half);
        return [axis[0] * s, axis[1] * s, axis[2] * s, Math.cos(half)];
    }

    _composePoses(p1, p2) {
        // p1 * p2
        const pos = this._addVec(p1.position, this._rotateVec(p1.orientation, p2.position));
        const ori = this._mulQuat(p1.orientation, p2.orientation);
        return { position: pos, orientation: ori };
    }

    _addVec(a, b) {
        return [a[0] + b[0], a[1] + b[1], a[2] + b[2]];
    }

    _rotateVec(q, v) {
        // Rotate vector v by quaternion q
        const qv = [v[0], v[1], v[2], 0];
        const qConj = [-q[0], -q[1], -q[2], q[3]];
        const result = this._mulQuat(this._mulQuat(q, qv), qConj);
        return [result[0], result[1], result[2]];
    }

    _mulQuat(a, b) {
        return [
            a[3]*b[0] + a[0]*b[3] + a[1]*b[2] - a[2]*b[1],
            a[3]*b[1] - a[0]*b[2] + a[1]*b[3] + a[2]*b[0],
            a[3]*b[2] + a[0]*b[1] - a[1]*b[0] + a[2]*b[3],
            a[3]*b[3] - a[0]*b[0] - a[1]*b[1] - a[2]*b[2]
        ];
    }

    clear() {
        console.log(`Clearing scene, removing ${Object.keys(this.nodes).length} tracked nodes`);

        // Remove all tracked nodes from rootGroup
        for (const [name, node] of Object.entries(this.nodes)) {
            this.rootGroup.remove(node.group);

            // Dispose geometries and materials to free memory
            if (node.mesh) {
                if (node.mesh.geometry) node.mesh.geometry.dispose();
                if (node.mesh.material) node.mesh.material.dispose();
            }
            console.log(`  Removed tracked: ${name}`);
        }

        // Also remove any stray Group objects from rootGroup (in case of race conditions)
        const toRemove = [];
        for (const child of this.rootGroup.children) {
            if (child.isGroup && child.name) {
                toRemove.push(child);
            }
        }
        for (const obj of toRemove) {
            console.log(`  Removed stray: ${obj.name}`);
            this.rootGroup.remove(obj);
        }

        this.nodes = {};
        this.treeData = null;
        this.jointCoords = {};

        console.log(`RootGroup children remaining: ${this.rootGroup.children.length}`);
    }
}

File: server/kinematic.py
"""
Kinematic Tree - Server-side calculations
"""

import math
import json
from dataclasses import dataclass, field
from typing import Optional, List, Dict


@dataclass
class Vec3:
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0

    def __add__(self, other: "Vec3") -> "Vec3":
        return Vec3(self.x + other.x, self.y + other.y, self.z + other.z)

    def __mul__(self, scalar: float) -> "Vec3":
        return Vec3(self.x * scalar, self.y * scalar, self.z * scalar)

    def to_list(self) -> list:
        return [self.x, self.y, self.z]

    @staticmethod
    def from_list(lst: list) -> "Vec3":
        return Vec3(lst[0], lst[1], lst[2]) if lst else Vec3()


@dataclass
class Quat:
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0
    w: float = 1.0

    def __mul__(self, other: "Quat") -> "Quat":
        return Quat(
            self.w * other.x + self.x * other.w + self.y * other.z - self.z * other.y,
            self.w * other.y - self.x * other.z + self.y * other.w + self.z * other.x,
            self.w * other.z + self.x * other.y - self.y * other.x + self.z * other.w,
            self.w * other.w - self.x * other.x - self.y * other.y - self.z * other.z,
        )

    def rotate_vec(self, v: Vec3) -> Vec3:
        """Rotate vector by quaternion"""
        qv = Quat(v.x, v.y, v.z, 0)
        conj = Quat(-self.x, -self.y, -self.z, self.w)
        result = self * qv * conj
        return Vec3(result.x, result.y, result.z)

    def to_list(self) -> list:
        return [self.x, self.y, self.z, self.w]

    @staticmethod
    def from_list(lst: list) -> "Quat":
        return Quat(lst[0], lst[1], lst[2], lst[3]) if lst else Quat()

    @staticmethod
    def from_axis_angle(axis: Vec3, angle: float) -> "Quat":
        """Create quaternion from axis and angle (radians)"""
        half = angle / 2
        s = math.sin(half)
        return Quat(axis.x * s, axis.y * s, axis.z * s, math.cos(half))


@dataclass
class Pose:
    position: Vec3 = field(default_factory=Vec3)
    orientation: Quat = field(default_factory=Quat)

    def __mul__(self, other: "Pose") -> "Pose":
        """Compose two poses: self * other"""
        return Pose(
            position=self.position + self.orientation.rotate_vec(other.position),
            orientation=self.orientation * other.orientation,
        )

    def to_dict(self) -> dict:
        return {
            "position": self.position.to_list(),
            "orientation": self.orientation.to_list(),
        }


class KinematicNode:
    def __init__(self, data: dict, parent: Optional["KinematicNode"] = None):
        self.name = data.get("name", "unnamed")
        self.type = data.get("type", "transform")
        self.parent = parent
        self.children: List["KinematicNode"] = []

        # Local pose (from JSON)
        pose_data = data.get("pose", {})
        self.local_pose = Pose(
            position=Vec3.from_list(pose_data.get("position", [0, 0, 0])),
            orientation=Quat.from_list(pose_data.get("orientation", [0, 0, 0, 1])),
        )

        # Axis for rotator/actuator
        axis_data = data.get("axis", [0, 0, 1])
        self.axis = Vec3.from_list(axis_data)

        # Current joint coordinate
        self.coord = 0.0

        # Model data (pass through to client)
        self.model = data.get("model", {})

        # Computed global pose
        self.global_pose = Pose()

        # Parse children
        for child_data in data.get("children", []):
            child = KinematicNode(child_data, self)
            self.children.append(child)

    def set_coord(self, value: float):
        """Set joint coordinate (radians for rotator, units for actuator)"""
        self.coord = value

    def get_joint_transform(self) -> Pose:
        """Get transform introduced by joint movement"""
        if self.type == "rotator":
            q = Quat.from_axis_angle(self.axis, self.coord)
            return Pose(orientation=q)
        elif self.type == "actuator":
            offset = self.axis * self.coord
            return Pose(position=offset)
        return Pose()

    def compute_global_poses(self, parent_pose: Pose = None):
        """Recursively compute global poses for all nodes"""
        if parent_pose is None:
            parent_pose = Pose()

        # Global = parent * local * joint
        # Joint transform is included so the node's visual moves/rotates with the joint
        self.global_pose = parent_pose * self.local_pose * self.get_joint_transform()

        for child in self.children:
            child.compute_global_poses(self.global_pose)

    def get_all_poses(self) -> dict:
        """Get dictionary of all node poses {name: pose}"""
        result = {self.name: self.global_pose.to_dict()}
        for child in self.children:
            result.update(child.get_all_poses())
        return result

    def get_scene_data(self) -> dict:
        """Get scene data for client (poses + model info)"""
        result = {
            self.name: {
                "pose": self.global_pose.to_dict(),
                "model": self.model,
            }
        }
        for child in self.children:
            result.update(child.get_scene_data())
        return result

    def find_by_name(self, name: str) -> Optional["KinematicNode"]:
        if self.name == name:
            return self
        for child in self.children:
            found = child.find_by_name(name)
            if found:
                return found
        return None

    def get_all_joints(self) -> List["KinematicNode"]:
        """Get all joint nodes (rotator/actuator)"""
        joints = []
        if self.type in ("rotator", "actuator"):
            joints.append(self)
        for child in self.children:
            joints.extend(child.get_all_joints())
        return joints


class KinematicTree:
    def __init__(self):
        self.root: Optional[KinematicNode] = None
        self.joints: Dict[str, KinematicNode] = {}

    def load(self, data: dict):
        """Load tree from JSON data"""
        self.root = KinematicNode(data)
        self.joints = {j.name: j for j in self.root.get_all_joints()}
        self.update()

    def set_joint_coords(self, coords: Dict[str, float]):
        """Set multiple joint coordinates"""
        for name, value in coords.items():
            if name in self.joints:
                self.joints[name].set_coord(value)

    def update(self):
        """Recompute all global poses"""
        if self.root:
            self.root.compute_global_poses()

    def get_scene_data(self) -> dict:
        """Get all scene data for rendering"""
        if self.root:
            return self.root.get_scene_data()
        return {}

    def get_joint_names(self) -> List[str]:
        """Get list of joint names"""
        return list(self.joints.keys())
